src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -189,8 +189,11 @@ static async ValueTask Await(int i, ValueTask vt, List<object> toDispose)
 
             }
 
-            if (IsRootScope)
+            if (IsRootScope && !RootProvider.IsDisposed())
             {
+                // If this ServiceProviderEngineScope instance is a root scope, disposing this instance will need to dispose the RootProvider too.
+                // Otherwise the RootProvider will never get disposed and will leak.
+                // Note, if the RootProvider get disposed first, it will automatically dispose all attached ServiceProviderEngineScope objects.
                 RootProvider.Dispose();
```

Comment:
Why cant this just noop internally?

Reply:
It is simpler. no? 

More code, more bugs :-)

Reply:
> Why cant this just noop internally?

By "just noop internally", you mean adding the `if` check in `ServiceProvider` Dispose?

Reply:
Yep.